src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,128 @@
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Linq;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using System;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Information about rename document calls that allows them to be applied as individual actions.
+        /// 
+        /// To apply all actions use <see cref="UpdateSolutionAsync(Solution, CancellationToken)"/>, or use a subset
+        /// of the actions by calling <see cref="UpdateSolutionAsync(Solution, ImmutableArray{RenameDocumentAction}, CancellationToken)"/>. 
+        /// Each action has a description of the changes that it will apply that can be presented to a user.
+        /// </summary>
+        public sealed class RenameDocumentActionSet
+        {
+            private readonly DocumentId _documentId;
+            private readonly string _documentName;
+            private readonly IReadOnlyList<string> _documentFolders;
+            private readonly ProjectId _projectId;
+            private readonly OptionSet _optionSet;
+
+            internal RenameDocumentActionSet(
+                ImmutableArray<RenameDocumentAction> actions,
+                ProjectId projectId,
+                DocumentId documentId,
+                string documentName,
+                IReadOnlyList<string> documentFolders,
+                OptionSet optionSet)
+            {
+                ApplicableActions = actions;
+                _documentFolders = documentFolders;
+                _documentId = documentId;
+                _documentName = documentName;
+                _projectId = projectId;
+                _optionSet = optionSet;
+            }
+
+            /// <summary>
+            /// All applicable actions computed for the action
+            /// </summary>
+            public ImmutableArray<RenameDocumentAction> ApplicableActions { get; }
+
+            /// <summary>
+            /// Same as calling <see cref="UpdateSolutionAsync(Solution, ImmutableArray{RenameDocumentAction}, CancellationToken)"/> with 
+            /// <see cref="ApplicableActions"/> as the argument
+            /// </summary>
+            public Task<Solution> UpdateSolutionAsync(Solution solution, CancellationToken cancellationToken)
+                => UpdateSolutionAsync(solution, ApplicableActions, cancellationToken);
+
+            /// <summary>
+            /// Applies each <see cref="RenameDocumentAction"/> in order and returns the final solution. 
+            /// All actions must be contained in <see cref="ApplicableActions" />
+            /// </summary>
+            /// <remarks>
+            /// An empty action set is still allowed and will return a modified solution
+            /// that will update the document properties as appropriate.
```

Comment:
is this a realistic use case?  can you flesh out when this would be appropriate/desired?

Reply:
Did my best to comment more...